overlord/state/copy.go

```diff
@@ -0,0 +1,140 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package state
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/snapcore/snapd/jsonutil"
+	"github.com/snapcore/snapd/osutil"
+)
+
+type checkpointOnlyBackend struct {
+	path string
+}
+
+func (b *checkpointOnlyBackend) Checkpoint(data []byte) error {
+	return osutil.AtomicWriteFile(b.path, data, 0600, 0)
+}
+
+func (b *checkpointOnlyBackend) EnsureBefore(d time.Duration) {
+	panic("cannot use EnsureBefore in checkpointOnlyBackend")
+}
+
+func (b *checkpointOnlyBackend) RequestRestart(t RestartType) {
+	panic("cannot use RequestRestart in checkpointOnlyBackend")
+}
+
+// copyData will copy the given subkeys specifier from srcData to dstData.
+//
+// The subkeys is constructed from a dotted path like "user.auth". This copy
+// helper is recursive and the pos parameter tells the function the current
+// position of the copy.
+func copyData(subkeys []string, pos int, srcData map[string]*json.RawMessage, dstData map[string]interface{}) error {
+	if pos < 0 || pos > len(subkeys) {
+		return fmt.Errorf("internal error: copyData used with an out-of-bounds position: %v not in [0:%v]", pos, len(subkeys))
+	}
+	raw, ok := srcData[subkeys[pos]]
+	if !ok {
+		return ErrNoState
+	}
+
+	if pos+1 == len(subkeys) {
```

Comment:
What if `len(subkeys) == 0`, initially `pos` will be `0` so this condition will not be met. Hmm. I think it's not a big deal (seeing how copyData is called below) but we should ensure that `len(subkeys) > 0`